code.py

```diff
@@ -1,27 +1,25 @@
 from sqlalchemy import create_engine
+from time import time
 import pandas as pd
-from multiprocessing import Pool, cpu_count
-import json
+import toml
 import os
 
-data = json.load(open('map.json'))
+start = time()
 
-for elem in data:
+config = toml.load(open('config.toml'))["config"]
+
+if not os.path.exists('parquet'): os.mkdir('parquet')
+
+for elem in config:
     if (elem["location"] == "local"):
         pass
     else:
         db_connection_str = f'mysql+pymysql://{elem["DB_USER"]}@{elem["location"]}:{elem["port"]}/{elem["species"]}'
         engine = create_engine(db_connection_str)
 
-        queryVars = {}
-        for i in elem["data"]["vars"]:
-            queryVars[i["key"]] = pd.read_sql(i["query"], con=engine)[i["key"]].item()
-
         for table in elem["data"]["tables"]:
             query = table["query"]
-            for queryVar in queryVars:
-                if ("{"+queryVar+"}" in query): query = query.replace("{"+queryVar+"}", str(queryVars[queryVar]))
-            df = pd.read_sql(query, con=engine)
+            df = pd.read_sql(query, con=engine, params= [table["vars"]])
 
             # re-assign column names if duplicate column names exist
             if ("columns" in table.keys()): df.columns = table["columns"].split(", ")
```

Comment:
Rather than doing this, we can instead rename the columns in the SQL itself using AS clauses for example

```sql
SELECT distinct g.gene_id as gene_id, g.stable_id as gene_stable_id, g.version as gene_stable_id_version, sr.name as chr, g.seq_region_start as gene_start, g.seq_region_end as gene_end, g.biotype as gene_biotype, x.dbprimary_acc as gene_symbol_id, x.display_label as gene_symbol, ct.stable_id as canonical_transcript_stable_id, ct.version as canonical_transcript_stable_id_version, ct.biotype as canonical_transcript_biotype
```

That way all the columns are now uniquely named.